### ISSUE TITLE
[CI][Packaging][Java] java-jars job fail on macOS aarch_64

### ISSUE THREAD

**Issue by raulcd** [apache/arrow#37201]
### Describe the bug, including details regarding any error messages, version, and platform.

The java-jars job is currently failing on [Build C++ libraries macOS aarch_64](https://github.com/ursacomputing/crossbow/actions/runs/5865193153/job/15901640364#logs)

This error is also happening on the maintenance branch for 13.0.0:
https://github.com/apache/arrow/pull/36616#issuecomment-1680342110
https://github.com/ursacomputing/crossbow/actions/runs/5877563237/job/15937924010

The error is the following:

```
 -- Performing Test CXX_LINKER_SUPPORTS_VERSION_SCRIPT
 -- Performing Test CXX_LINKER_SUPPORTS_VERSION_SCRIPT - Failed
```
There's a bunch of Warnings but I am pretty sure they are not related:
```
CMake Warning at /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp-install/lib/cmake/Arrow/FindThriftAlt.cmake:56 (find_package):
  By not providing "FindThrift.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Thrift", but
  CMake did not find one.

  Could not find a package configuration file provided by "Thrift" with any
  of the following names:

    ThriftConfig.cmake
    thrift-config.cmake

  Add the installation prefix of "Thrift" to CMAKE_PREFIX_PATH or set
  "Thrift_DIR" to a directory containing one of the above files.  If "Thrift"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.27.2/share/cmake/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp-install/lib/cmake/Arrow/ArrowConfig.cmake:105 (find_dependency)
  adapter/orc/CMakeLists.txt:18 (find_package)


CMake Warning at /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp-install/lib/cmake/Arrow/Findlz4Alt.cmake:29 (find_package):
  By not providing "Findlz4.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "lz4", but
  CMake did not find one.

  Could not find a package configuration file provided by "lz4" with any of
  the following names:

    lz4Config.cmake
    lz4-config.cmake

  Add the installation prefix of "lz4" to CMAKE_PREFIX_PATH or set "lz4_DIR"
  to a directory containing one of the above files.  If "lz4" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.27.2/share/cmake/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp-install/lib/cmake/Arrow/ArrowConfig.cmake:105 (find_dependency)
  adapter/orc/CMakeLists.txt:18 (find_package)


-- Arrow version: 13.0.0
-- Found the Arrow shared library: 
-- Found the Arrow import library: 
-- Found the Arrow static library: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/arrow/cpp-build/cpp-install/lib/libarrow.a
-- Configuring incomplete, errors occurred!
```

### Component(s)

Continuous Integration, Java, Packaging

**Comment by raulcd** [comment]
It seems the error is here:
```
 xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
CMake Error at /opt/homebrew/Cellar/cmake/3.27.2/share/cmake/Modules/FindJNI.cmake:394 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.
```

**Comment by raulcd** [comment]
I am investigating with @amaldonadomat it might be related to an update on the M1 OS

**Comment by raulcd** [comment]
It does seem to be related to the last two commits on `FindJNI.cmake`:
https://github.com/Kitware/CMake/commit/efa0e2ddafe48ef0921b31aef37c3187a7bb6b97
https://github.com/Kitware/CMake/commit/597139b853f2f14c71047e41a3244a397c6ee728

**Comment by raulcd** [comment]
it does seem the fix is not included on a new CMake release

**Comment by raulcd** [comment]
There doesn't seem to be any alias for prior CMake versions on Homebrew that we could reuse.

**Comment by raulcd** [comment]
I am asking if they have a timeline for 3.27.3 on the fix: https://gitlab.kitware.com/cmake/cmake/-/issues/25181

**Comment by assignUser** [comment]
This is on our self-hosted runner right? I think the easiest fix would be to manually install an [older cmake version](https://github.com/Kitware/CMake/releases/download/v3.26.5/cmake-3.26.5-macos-universal.dmg) without the bug and remove the brew cmake after the bundle installation: https://github.com/ursacomputing/crossbow/actions/runs/5877563237/workflow#L197

**Comment by kou** [comment]
Duplicated of #37159.

**Comment by kou** [comment]
How about adding `-DXCODE_VERSION=` CMake option as a workaround?

```diff
diff --git a/ci/scripts/java_jni_macos_build.sh b/ci/scripts/java_jni_macos_build.sh
index d66c39a37..647a7dad8 100755
--- a/ci/scripts/java_jni_macos_build.sh
+++ b/ci/scripts/java_jni_macos_build.sh
@@ -90,6 +90,7 @@ cmake \
   -DPARQUET_BUILD_EXECUTABLES=OFF \
   -DPARQUET_REQUIRE_ENCRYPTION=OFF \
   -Dre2_SOURCE=BUNDLED \
+  -DXCODE_VERSION= \
   -GNinja \
   ${arrow_dir}/cpp
 cmake --build . --target install
```

**Comment by kou** [comment]
https://gitlab.kitware.com/cmake/cmake/-/issues/25181#note_1401407

> I managed to wrap up that work quickly. I've released 3.27.3. It should be announced later today.

https://github.com/Homebrew/homebrew-core/pull/139719

We may not need a workaround.
